Support cross-drive linking on windows

cspotcode opened the pull request:
Fixes #17, duplicates #16

In order to test cross-drive linking, I enabled dependency injection for the `fs` dependency.  The unit tests inject an in-memory fs module.  This allows simulating multiple drives on Windows.

~Might not be 100% ready, but creating this PR allows unit tests to run on CI, which saves me time.~

cspotcode committed:
Support cross-drive linking on windows

cspotcode committed:
remove mz dependency

cspotcode commented:
This is ready for review.

zkochan commented on index.js:
```diff
@@ -34,7 +35,22 @@ cmdShim.ifExists = cmdShimIfExists
+let fs
+let fsRaw
+setFs(require('fs'))
+function setFs (_fs) {
```
why don't you just pass in fs via an optional `opts.fs`?

cspotcode replied:
I was avoiding changes to the API.  But if you're ok adding the new option, that works too.

zkochan replied:
yeah, that seems like a widespread convention. Let's add this via an option

cspotcode committed:
Allow new option "fs" to set filesystem implementation to use.

Since this affects the API, and I didn't want to duplicate the changes in 2 places, I made things DRY by porting to TypeScript.
index.d.ts and index.js are both created from the compile step.

cspotcode committed:
Renaming index.js to src/index.ts

cspotcode commented:
Updated to add the `fs` option which allows the caller to pass an `fs` implementation to be used.

Since this involves and API change, and I don't like updating .d.ts and .js in parallel, I went ahead and ported to .ts.

To see a diff of the source code changes, plus the changes to the runtime .js code and to the .d.ts 
which is now being emitted by the compiler instead of written manually, you can look here:
https://github.com/cspotcode/cmd-shim/compare/cross-drive-links-diff-demonstration~1..cross-drive-links-diff-demonstration

cspotcode committed:
better logging from tsc

cspotcode committed:
Add clean script

ExE-Boss commented:
I think that the conversion to TypeScript should be done in a separate PR.

zkochan commented on src/index.ts:
```diff
@@ -440,25 +507,21 @@ function generatePwshShim (src, to, opts) {
   }
   let split = (typeof nodePath === 'string' ? nodePath.split(path.delimiter) : Array.from(nodePath))
-  let result = {}
+  let result = {} as {win32:string,posix:string,[index:number]:{win32:string,posix:string}}
```
I think you can do it like

```ts
let result = {} as ReturnType<typeof normalizePathEnvVar>
```

zkochan commented on test/basic.js:
```diff
@@ -431,3 +431,65 @@ test('explicit shebang with args', function (t) {
 })
+
+if (process.platform === 'win32') {
+  test('explicit shebang with args, linking to another drive on Windows', function (t) {
```
maybe it would be nicer w/o the nesting. something like

```ts
const testWindows = process.platform === 'win32' ? test : test.skip

testWindows('explicit shebang with args, linking to another drive on Windows', function (t) {
```

zkochan commented on test/00-setup.js:
```diff
@@ -2,31 +2,43 @@
-  'src.env.args': '#!/usr/bin/env node --expose_gc\ngc()\n',
-  'src.sh': '#!/usr/bin/sh\necho hi\n',
-  'src.sh.args': '#!/usr/bin/sh -x\necho hi\n'
+const {fixtures, fixtures2} = process.platform === 'win32' ? {
```
`is-windows` is in the deps, so you may use that

cspotcode replied:
Should I use `is-windows` in this case?  The conditional here is equivalent to testing if `path.win32 === path`, which will be true on Windows but not on Cygwin, whereas I think `is-windows` returns true for Cygwin.

What I'm really trying to test here is not if we're on Windows but if we're running a Windows build of nodejs, since that affects which `path` implementation is used by default.

On Linux, `path.relative()` can always return a relative path, even if it means `..`ing to the root of the filesystem.  On Windows, `path.relative()` can't do that when the drive is different, so we're able to call `path.relative()` and then check if the result `isAbsolute()`.

Does cmd-shim need to create .cmd shims from within a Linux build of nodejs?  (E.g. on WSL or Cygwin)  If so, I'll need to tweak the code to special-case this.  To properly support WSL, we'll need to call the external `wslpath` binary to convert the paths to Windows, then use `path.win32` to manipulate them.

zkochan replied:
I don't know the answer to this. I think the path.relative tweaks can be done in a separate PR. As far as I understand, this PR does not change how path.relative is used

cspotcode replied:
Without those tweaks, this PR won't be doing its job of allowing cross-drive linking.  Here's an example:

If I want to link from C:\foo to D:\bar, and I run it on Windows, `path.relative('C:\\', 'D:\\bar')` will return 'D:\bar'.  `path.isAbsolute()` will return true, and the %~dp0 prefix will be omitted from the path in the shim.

If I do the same in Cygwin or WSL, `path.relative('/c', '/d/bar')` will return '../d/bar'.  `path.isAbsolute()` will return false, and the `%~dp0` prefix will be included in the shim, so the shim will fail on Windows.  `C:\..\d\bar` is not what we want.

I think the question is: do we want to support creating cross-drive links from a Linux build of node running in WSL and/or Cygwin?  If "no", then this PR does the trick already.  If "yes", we have some more work to do.

zkochan replied:
Currently, it doesn't work on Windows at all, so this is already a huge improvement, as it is

zkochan commented on package.json:
```diff
@@ -8,32 +8,41 @@
+    "index.js.map"
   ],
-  "types": "index.d.ts",
+  "main": "index",
```
this is not needed

zkochan commented on src/index.ts:
```diff
@@ -57,133 +111,146 @@ const extensionToProgramMap = new Map([
 ])
 
+function ingestOptions (opts: Options): InternalOptions {
+  const opts_ = Object.assign({}, DEFAULT_OPTIONS, opts) as InternalOptions
```
you can use the spread operator here

zkochan reviewed (approved):
overall, LGTM

zkochan commented:
if not too hard, the conversion in a separate PR would be nice

zkochan commented on tsconfig.json:
```diff
@@ -0,0 +1,67 @@
+
+    /* Experimental Options */
+    // "experimentalDecorators": true,        /* Enables experimental support for ES7 decorators. */
+    // "emitDecoratorMetadata": true,         /* Enables experimental support for emitting type metadata for decorators. */
```
remove the commented out fields

cspotcode committed:
Addressing PR feedback

zkochan committed:
Update basic.js
